core/rawdb/bor_receipts.go

```diff
@@ -0,0 +1,211 @@
+package rawdb
+
+import (
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ledgerwatch/erigon-lib/kv"
+	"github.com/ledgerwatch/erigon/common"
+	"github.com/ledgerwatch/erigon/core/types"
+	"github.com/ledgerwatch/erigon/rlp"
+	"github.com/ledgerwatch/log/v3"
+)
+
+var (
+	// bor receipt key
+	borReceiptKey = types.BorReceiptKey
+)
+
+func borTxLookupKey(hash common.Hash) []byte {
+	return append(hash.Bytes(), hash.Bytes()...)
+}
+
+// HasBorReceipt verifies the existence of all block receipt belonging
+// to a block.
+func HasBorReceipts(db kv.Has, hash common.Hash, number uint64) bool {
+	if has, err := db.Has(kv.BorReceipts, borReceiptKey(number)); !has || err != nil {
+		return false
+	}
+	return true
+}
+
+// ReadBorReceiptRLP retrieves the block receipt belonging to a block in RLP encoding.
+func ReadBorReceiptRLP(db kv.Getter, hash common.Hash, number uint64) rlp.RawValue {
+	data, err := db.GetOne(kv.BorReceipts, borReceiptKey(number))
+	if err != nil {
+		log.Error("ReadBorReceiptRLP failed", "err", err)
+	}
+	return data
+}
+
+// ReadRawBorReceipt retrieves the block receipt belonging to a block.
+// The receipt metadata fields are not guaranteed to be populated, so they
+// should not be used. Use ReadBorReceipt instead if the metadata is needed.
+func ReadRawBorReceipt(db kv.Tx, hash common.Hash, number uint64) *types.Receipt {
+	// Retrieve the flattened receipt slice
+	data := ReadBorReceiptRLP(db, hash, number)
+	if data == nil || len(data) == 0 {
+		return nil
+	}
+
+	// Convert the receipts from their storage form to their internal representation
+	var storageReceipt types.ReceiptForStorage
+	if err := rlp.DecodeBytes(data, &storageReceipt); err != nil {
+		log.Error("Invalid receipt array RLP", "hash", hash, "err", err)
+		return nil
+	}
+
+	return (*types.Receipt)(&storageReceipt)
+}
+
+// ReadBorReceipt retrieves all the bor block receipts belonging to a block, including
+// its correspoinding metadata fields. If it is unable to populate these metadata
+// fields then nil is returned.
+func ReadBorReceipt(db kv.Tx, hash common.Hash, number uint64) *types.Receipt {
+	// We're deriving many fields from the block body, retrieve beside the receipt
+	borReceipt := ReadRawBorReceipt(db, hash, number)
+	if borReceipt == nil {
+		return nil
+	}
+
+	// We're deriving many fields from the block body, retrieve beside the receipt
+	receipts := ReadRawReceipts(db, number)
+	if receipts == nil {
+		receipts = make(types.Receipts, 0)
+	}
+
+	body, _, _ := ReadBody(db, hash, number)
```

Comment:
Getting entire body just to check it exists is expensive. Can just check 1 DB record in `BlockBody` table